packages/indicators/src/getExpressionParserInstance.ts

```diff
@@ -0,0 +1,23 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { ExpressionParser } from '@tupaia/expression-parser';
+
+let parserInstance: ExpressionParser | null = null;
+
+/**
+ * Creating a new `ExpressionParser` instance takes around 20-30ms. This can have a significant
+ * performance impact when building data for nested indicators, where ExpressionParser would
+ * be independently instantiated multiple times.
+ *
+ * We can safely use a Singleton pattern since the creation of ExpressionParser is not coupled to
+ * specific app data/logic
+ */
+export const getExpressionParserInstance = () => {
+  if (!parserInstance) {
+    parserInstance = new ExpressionParser();
+  }
+  return parserInstance;
+};
```

Comment:
I think this is good for now, but perhaps in the future we may want to implement and singleton pattern on the expression-parser library itself? For instance the report-server creates instances of it for each transform, and an instances is being created here whenever the report-server queries for indicators... I think it would be optimal to have one shared instances across the whole process?

Reply:
I thought of that too, and initially when with that. It reads even nice (`ExpressionParser.getSingleton()`)

But then I thought it is a bit risky to share state management between packages. Eg ExpressionParser.setScope will change the state of the instance, and the client will have to explicitly call `clearScope` afterwards. Or some clients may choose to dynamically add some functions to ExpressionParser (I believe math.js makes this possible). It would be harder to track things if another client in another package messes up with your instance.

And since the initialisation is 20-30ms, doing it once per involved package still shouldn't add too much overhead overall (especially given this will happen once per package + per the lifetime of the node process). And it seems a bit easier to track programmatically. What do you think?

Reply:
Yeah I'm kind of on the fence about it as well... I'm definitely happy with doing this per-package for now 👍